python/dgl/graphbolt/item_sampler.py

```diff
@@ -326,8 +329,114 @@ def __iter__(self) -> Iterator:
+    (process) receives an exclusive subset.
+
+    Note: This class `DistributedItemSampler` is not decorated with
+    `torchdata.datapipes.functional_datapipe` on purpose. This indicates it
+    does not support function-like call. But any iterable datapipes from
+    `torchdata` can be further appended.
+
+    Parameters
+    ----------
+    item_set : ItemSet or ItemSetDict
+        Data to be sampled.
+    batch_size : int
+        The size of each batch.
+    minibatcher : Optional[Callable]
+        A callable that takes in a list of items and returns a `MiniBatch`.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    shuffle : bool
+        Option to shuffle before sample.
+    num_replicas: int
+        The number of model replicas that will be created during Distributed
+        Data Parallel (DDP) training. By default, the world size is retrieved
+        from the current distributed group.
+    drop_uneven_inputs : bool
+        Option to make sure the numbers of batches for each replica are the
+        same. If some of the replicas have more batches than the others, the
+        redundant batches of those replicas will be dropped. If the drop_last
+        parameter is also set to True, the last batch will be dropped before the
+        redundant batches are dropped.
+        Note: When using Distributed Data Parallel (DDP) training, the program
+        may hang or error if the a replica has fewer inputs. It is recommended
+        to use the Join Context Manager provided by PyTorch to solve this
+        problem. Please refer to
+        https://pytorch.org/tutorials/advanced/generic_join.html. However, this
+        option can be used if the Join Context Manager is not helpful for any
+        reason.
+
+    Examples
+    --------
+    """
+
+    def __init__(
+        self,
+        item_set: ItemSet or ItemSetDict,
+        batch_size: int,
+        minibatcher: Optional[Callable] = minibatcher_default,
+        drop_last: Optional[bool] = False,
+        shuffle: Optional[bool] = False,
+        num_replicas: Optional[int] = None,
+        drop_uneven_inputs: Optional[bool] = False,
+    ) -> None:
+        super().__init__(item_set, batch_size, minibatcher, drop_last, shuffle)
+        self._drop_uneven_inputs = drop_uneven_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
+        # Get world size.
+        if num_replicas is None:
+            assert (
+                dist.is_available()
+            ), "Requires distributed package to be available"
+            num_replicas = dist.get_world_size()
+        total_len = len(item_set)
+        # Calculate the number of batches for each replica.
+        self._num_batches = total_len // (num_replicas * batch_size) + (
+            (not drop_last)
+            and (total_len % (num_replicas * batch_size) >= num_replicas)
         )
 
+    def __iter__(self) -> Iterator:
```

Comment:
below code is quite like to the implementation of `ItemSampler`. is it possible to remove such code duplication? In current way, I don't see much benefit of subclassing `ItemSampler`.